Tighten tolerance of test_vmapvjp_linalg_tensorsolve_cpu_float32

Flamefire opened the pull request:
With the optimzation of `solve` using a transposed input this test fails to meet these tolerances but passes without.

@pytorchbot topic: not user facing

Flamefire committed:
Tighten tolerance of test_vmapvjp_linalg_tensorsolve_cpu_float32

With the optimzation of `solve` using a transposed input this test
fails to meet these tolerances but passes without.

zou3519 reviewed (changes requested):
tests failing?

Flamefire commented:
> tests failing?

This is an intended reproducer for a bug. See https://github.com/pytorch/pytorch/pull/151896#issuecomment-2830442130

The open question is if we should tighten the tolerances to zero which results in effectively using the default tolerances for the floating point type, see https://github.com/pytorch/pytorch/pull/151896#discussion_r2060363330
I'd say the current ones are enough but am waiting for a reply from @lezcano what he prefers

lezcano commented:
I already answered there.

The tl;dr is that either we always use exactly the same codepath for every device in both computations and we can set the atol=rtol=0, or we don't. As I said there, it's clearly the case we don't, so then  see https://github.com/pytorch/pytorch/pull/151896#discussion_r2060314684

lezcano commented:
why is the test showing a 10^-5 difference if the same algorithm is executed?

lezcano commented:
anyway, I really think we should drop this PR and land the other one without a regression test

Flamefire commented:
> why is the test showing a 10^-5 difference if the **exact** same algorithm is executed?

Because prior to that PR it was not using the exact same algorithm: At some point it transposed the input in one case.   
With the PR it should be the same now.

> anyway, I really think we should drop this PR and land the other one without a regression test

If you are sure I'd revert the tolerance change commit but keep the commit so that it is still in the history, especially with the CI status, for future reference. Or would you rather like me to fully remove that commit?

lezcano commented:
the commits are squashed so it doesn't really matter

Flamefire committed:
Reduce tolerance for linalg.tensorsolve to zero

The actual tolerance will be increased to the minimum tolerance for the
float32 type in the test code but this shows the intention as explained
by the comment.

Flamefire commented:
Alright, thanks!

I removed this change from #151896 and updated this PR to request a tolerance of zero  as [you requested in the other PR](https://github.com/pytorch/pytorch/pull/151896#discussion_r2060363330). A higher one will be used which I shortly described in the commit message and comment, just to avoid confusion.

> If the two are actually calling exactly the same ops and we can guarantee bitwise equivalence, then let's set atol = rtol = 0.

I'm convinced this is true for the CPU algorithm as explained above, but cannot tell for sure for the GPU/XPU code so I restricted it to CPU.
I also added an explanation to the other PR description that it makes the algorithms identical (for now)

If this test change now makes sense I can merge it to #151896 to show it passes with that change. Otherwise that PR can be merged without a reproducer and we can close this one.

lezcano commented:
As discussed in https://github.com/pytorch/pytorch/pull/152180#issuecomment-2838997180 and https://github.com/pytorch/pytorch/pull/152180#issuecomment-2839437702, the algorithms are not the same for all backends, and even if they are now for CPU, you cannot be sure that this will be the case in the future.

Let's please drop this comit.

Flamefire commented:
Sure, just wanted to provide a test for the current behavior and that detects if this (the CPU algo) is changed in a way that causes the failure unintended.
Closing this PR then. The other already contains just the fix already so can be merged.
